src/MemoryTracing/include/MemoryTracing/MemoryInfoProducer.h

```diff
@@ -28,17 +26,16 @@ class MemoryInfoProducer {
   MemoryInfoProducer(MemoryInfoProducer&&) = delete;
   MemoryInfoProducer& operator=(MemoryInfoProducer&&) = delete;
 
-  ~MemoryInfoProducer() { Stop(); }
+  virtual ~MemoryInfoProducer() { Stop(); }
 
   void SetListener(MemoryInfoListener* listener) { listener_ = listener; }
 
   void Start();
   void Stop();
 
- private:
+ protected:
   void SetExitRequested(bool exit_requested);
-  void Run();
-  void ProduceSystemMemoryUsageAndSendToListener();
+  virtual void Run() = 0;
```

Comment:
I think I would extract this to a separate interface or even just a functor maybe, and give it in constructor to this class. add pid and the listener as parameters this way you can keep them private.

Reply:
Thanks for the suggestion! I changed to create system, cgroup and process memory information producers with different run functions.